BENCH: optimize: add DFO CUTEST benchmark

sarafridov opened the pull request:
#### What does this implement/fix?
This PR adds the smooth DFO benchmark set of Moré and Wild, and changes the default behavior of Wolfe line search (when called from BFGS) to use cheaper directional derivative finite differencing if the gradient function is based on finite differencing (if a regular gradient function is provided, that is still used). 

#### Additional information
The DFO benchmark results on the current scipy (for BFGS):
<img width="637" alt="DFO benchmark problem number" src="https://user-images.githubusercontent.com/16036159/129147505-11c3b7e2-ed65-4a17-b6d6-bc97c2096c98.png">

The DFO benchmark results on the version in this PR (for BFGS):
<img width="623" alt="DFO benchmark problem number" src="https://user-images.githubusercontent.com/16036159/129147585-6037b76d-2e5e-4ea5-90d1-73d300646f61.png">

These are summarized in a histogram of performance ratio (number of function evaluations used by each version divided by the number of function evaluations used by the fastest of the two versions): 
![image](https://user-images.githubusercontent.com/16036159/129152236-aa919e74-1cc1-4626-98a2-54a3c4f3f4d9.png)
For most problems either approach is about the same (within a factor of 2), but there are a few problems where the new version is much faster.

tylerjereddy commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -202,6 +202,40 @@ def test_bfgs(self):
+        
+        # Ensure that function call counts are 'known good'; these are from
+        # SciPy 1.21.1. Don't allow them to increase.
+        assert_(self.funccalls == 58, self.funccalls) 
```
some of the CI failures look related:

```
scipy/optimize/tests/test_optimize.py:230: in test_bfgs_dfo
    assert_(self.funccalls == 58, self.funccalls)
E   AssertionError: 48
```

sarafridov replied:
What command did you run to get this error? When I run `python runtests.py -s optimize` all the tests (including this one) pass.

sarafridov replied:
Oh I see it in the continuous integration tests...but I'm not sure why the behavior is different there vs on my machine?

rgommers replied:
It's not that uncommon for there to be platform-specific differences, maybe due to floating point behavior, use of a different compiler, BLAS library, etc.

These kind of function call checks are fragily, and in this case getting a lower number of function calls is a good thing. I think changing `== 58` to `<= 58` should be fine.

tylerjereddy commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -202,6 +202,40 @@ def test_bfgs(self):
+        # Ensure that function call counts are 'known good'; these are from
+        # SciPy 1.21.1. Don't allow them to increase.
+        assert_(self.funccalls == 58, self.funccalls) 
+        assert_(self.gradcalls == 0, self.gradcalls) 
```
I think we have a modest preference for using plain `assert` with `pytest` these days

sarafridov replied:
I used `assert_` just to be consistent with the rest of the file, but I see the discussion on preferring `assert` in new code...would it be helpful for me to change that to `assert` for the whole file, or just my added test?

sarafridov replied:
I just made a separate PR to replace `assert_` with `assert` in `test_optimize.py` and `test_linesearch.py`.

rgommers replied:
Thanks, that PR is merged now.

tylerjereddy commented on benchmarks/benchmarks/optimize.py:
```diff
@@ -262,6 +264,48 @@ def bench_run(self, x0, methods=None, **minimizer_kwargs):
+    ]
+    param_names = ["DFO benchmark problem number", "solver", "result type"]
+
+    def setup(self, prob_number, method_name, ret_val):
```
Not really related to this PR since I see that the rest of this module is built the same way as well, but this is certainly quite different from the normal structure of conventional `asv` benchmarks.

I assume there's a good reason for having a custom `bench_run()` throughout this file. Out of curiosity, does `asv continuous master my_branch` even work for assessing performance improvements in `optimize`? The PR description here goes through a different mechanism to compare for example.

sarafridov replied:
This is my first time working with `asv`, so I modeled my benchmark on the rest of the module. However, when I try to compare branches (to benchmark this PR) with `asv` it gets stuck and times out creating the environments. So, I ended up just making my own local branches with the old and new versions, running the benchmark in each branch, and manually comparing the results.

andyfaff commented:
I guess another alternative that could be tried is a wrapper function?

tylerjereddy commented:
@sarafridov It looks like there are still a few reviewer comments to address. Since we'll be cutting the next release branch soon I may bump the milestone if you don't have the bandwidth to circle back here before Monday.

sarafridov commented:
Yes, sorry I got busy with other stuff but I’ve been meaning to finish this. I’ll try to make the edits and reply before Monday, but if I don’t have time it’s ok to put it in the next release.

Thanks,
Sara

> On Dec 2, 2021, at 7:32 PM, Tyler Reddy ***@***.***> wrote:
> 
> 
> @sarafridov <https://github.com/sarafridov> It looks like there are still a few reviewer comments to address. Since we'll be cutting the next release branch soon I may bump the milestone if you don't have the bandwidth to circle back here before Monday.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/scipy/scipy/pull/14579#issuecomment-985186711>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AD2LCP6AEJTCQLHJZ44U6HLUPA243ANCNFSM5CAM7WFA>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

sarafridov commented:
> 

I think I've addressed all of these issues, but there are some merge conflicts I'm not sure how to resolve (and these seem to be blocking the tests from running).

tylerjereddy reviewed (commented):
We should check if @andyfaff is "ok" with this PR now? I added one more comment; I also spent a bunch of time looking over formatting-only changes unfortunately, but I see there was a reviewer request to run a linter on a bunch of files.

From a "safety" standpoint, there are almost no changes to old tests, which makes this a bit less risky perhaps.

For the matter of rebasing, it is quite tricky because of our recent namespace privatization efforts. I was able to do it in feature branch `treddy_gh_14579_rebase`, which I've pushed up to my fork.

It took quite some time/effort to do that, but it seems to be a perfect match to the diff stats in this PR now, with the underscore prefix files also matching exactly to their previous state:

![image](https://user-images.githubusercontent.com/7903078/144732527-23273e92-d80f-4bfe-861f-fd7d2c540660.png)

tylerjereddy commented:
You may also need help figuring out how to push my branch into here. After spot checking that branch form my fork, and perhaps backing up your branch locally, you could so something similar to this to push from one branch to another with a different name (and would likely need a force push in this case with `-f`).

andyfaff commented:
I won't be able to look at this until mid next week.

sarafridov commented:
> 



> 

Thanks for fixing the merge conflict! I think I managed to get your branch into my local state, but I'm trying to run the tests locally before I push it back here, and I'm getting a build error `FileNotFoundError: [Errno 2] No such file or directory: 'scipy/stats/biasedurn.pxd'` when I run `python runtests.py`. I think this is happening because `biasedurn.pxd` was renamed, but I'm not sure why the test build seems to be expecting the old version. Are you able to run tests locally in your branch `treddy_gh_14579_rebase`?

sarafridov commented on benchmarks/benchmarks/optimize.py:
```diff
@@ -225,59 +245,192 @@ def bench_run(self, x0, methods=None, **minimizer_kwargs):
+            50,
+            51,
+            52,
+        ],
```
Done.

tylerjereddy commented:
I'll bump the milestone here because this is a big PR and I'm a bit nervous to merge it before branching without a high degree of certainty that it is "ok." It looks like I was supposed to help move this forward a bit more so I'll make a note to circle back soon if an `optimize` expert doesn't beat me to it.

sarafridov commented:
Thanks Tyler!

> On May 29, 2022, at 1:51 PM, Tyler Reddy ***@***.***> wrote:
> 
> 
> I'll bump the milestone here because this is a big PR and I'm a bit nervous to merge it before branching without a high degree of certainty that it is "ok." It looks like I was supposed to help move this forward a bit more so I'll make a note to circle back soon if an optimize expert doesn't beat me to it.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/scipy/scipy/pull/14579#issuecomment-1140522333>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AD2LCPYCKVLCQNZJOMRRUNTVMPKEXANCNFSM5CAM7WFA>.
> You are receiving this because you were mentioned.
>

mdhaber commented on scipy/optimize/_linesearch.py:
```diff
@@ -327,6 +386,9 @@ def scalar_search_wolfe2(phi, derphi, phi0=None,
         Objective scalar function.
     derphi : callable phi'(alpha)
         Objective function derivative. Returns a scalar.
+        derphi also accepts an optional argument fs, the 
```
```suggestion
        `derphi` also accepts an optional argument fs, the 
```

mdhaber commented:
@sarafridov Thanks for you patience! I'd like to help get this merged. One thing that I think would help is if we could split the benchmarks from the rest of the PR. Would you mind if we start with the benchmarks in this PR, then we can create a new PR for the line search enhancement? I can resolve the merge conflicts for you and push back to your branch, if that's OK with you. The line search enhancements will still be in the commit history, and I'll create a backup before merging just in case, but you might want to create your own backup branch, too.

sarafridov commented:
Sure, thanks Matt! Any help merging is appreciated :)

mdhaber commented:
Excellent I'll do that shortly!

mdhaber commented:
OK. I removed the directional derivative part, which we'll merge as part of another PR.
<details>
I thought about keeping _that_ part here and asking you to make the new PR with the benchmarks (instead of vice-versa) to preserve the link between comments and code, but there weren't many comments directly linked to the code anyway, and we can always refer back to the comments in this PR if need be. When the history gets long, sometimes it's nice to start with a clean slate.
</details>

A few questions about the benchmarks:
* Did [this commit](https://github.com/scipy/scipy/pull/14579/commits/3c7d22148e9690007c51cb49f6cbc38c12e34b97) make any non-formatting changes?  It looks like you used Black to format those. That's fine for new files, but please don't use it on existing files. SciPy hasn't adopted Black yet, so it makes it difficult to see what's really going on with amidst all the formatting changes.
* Are reference values for the optimal value of the objective function available? I'd like to track the difference between the values found by our optimizers and the reference value.
* Can you find the licenses for the references for any code you referenced for the benchmarks and link to them here? (It looks like you manually translated from Matlab to Python?)

Also, I see that Tyler [mentioned](https://github.com/scipy/scipy/pull/14579#pullrequestreview-729201337) that these are written like others in this file, but they don't follow the usual ASV pattern. If I were to provide a different reference to follow, would you be willing to refactor these, or would you prefer that I take care of it and push the changes here?

Thanks!

sarafridov commented:
* I think the only non-formatting change there was in `test_linesearch.py`, where it looks like I reverted a function signature change that I had added in a previous commit
* I think the optimal objective values are zero for all of these, but now I can't find where I got that from unfortunately.
* I did indeed manually translate from MATLAB to Python, since I couldn't find an existing Python implementation. The original MATLAB versions came from https://www.mcs.anl.gov/~more/dfo/, which is the supplement to the paper `Benchmarking Derivative-Free Optimization Algorithms by J. Moré and S. Wild. SIAM J. Optimization, Vol. 20 (1), pp.172-191, 2009`. Each MATLAB file has a header comment that references the paper and original authors but doesn't explicitly include a license as far as I can tell. However, the website says that the benchmark functions were derived from those in the CUTEr problem set, which includes this license: https://github.com/ralna/CUTEst/blob/master/LICENSE

Regarding reformatting, I probably won't have time to do much before mid-November; if you want to reformat things earlier please go for it, and otherwise I can take a stab at it then. Thanks!

mdhaber commented:
@rgommers @tylerjereddy thought I should check with you since you commented about the benchmarks above:

On one hand, the original CUTEst benchmarks are LGPL, and the Matlab translation from which these Python versions were derived are not explicitly licensed. On the other hand, there are only so many ways to write a mathematical function like these, and I don't think LGPL covers the mathematical functions themselves. What would you suggest to avoid copyright issues here? My inclination would be the conservative route - check with both the Matlab and the original CUTEst authors for permission to relicense these under BSD before including these.

rgommers commented:
> OK. I removed the directional derivative part, which we'll merge as part of another PR.

Would you mind updating the title and main PR description? I was confused for a while when looking at the PR diff.

> What would you suggest to avoid copyright issues here? My inclination would be the conservative route - check with both the Matlab and the original CUTEst authors for permission to relicense these under BSD before including these.

Agreed. In particular the CUTEst author permission seems important. There's enough code and it's specific enough that it's clear that it is derived from the original code; writing it from scratch probably would not have yielded such a close match.

mdhaber commented:
@sarafridov The CUTEst license is LGPL, and my understanding is that derivative works of LGPL code must be LGPL-licensed. This means that neither the Matlab code you referenced or the code you wrote can be distributed under a more permissive license, like SciPy's BSD 3-Clause license, without an exception being made. Would you be willing to contact the original authors of these benchmarks and ask for permission for SciPy to distribute them under our [BSD 3-Clause license](https://github.com/scipy/scipy/blob/main/LICENSE.txt)?

wildsm commented:
> @sarafridov The CUTEst license is [LGPL](https://fossa.com/blog/open-source-software-licenses-101-lgpl-license/#:~:text=And%20like%20the%20others%2C%20it's,the%20MIT%20or%20BSD%20licenses.), and my understanding is that derivative works of LGPL code must be LGPL-licensed. This means that neither the Matlab code you referenced or the code you wrote can be distributed under a more permissive license, like SciPy's BSD 3-Clause license, without an exception being made. Would you be willing to contact the original authors of these benchmarks and ask for permission for SciPy to distribute them under our [BSD 3-Clause license](https://github.com/scipy/scipy/blob/main/LICENSE.txt)?

@sarafridov can respond in full, but the original matlab (& fortran) code for this test set is BSD 3-slaused: https://github.com/POptUS/BenDFO

mdhaber commented:
Thanks @wildsm. It says in the paper that "We use a selection of problems from the CUTEr collection". Just to be explicit, this means that the code in [POptUS/BenDFO](https://github.com/POptUS/BenDFO) was written independently of any LGPL-licensed code; it was just written based on the same mathematical problems? If so, then yes, I don't think we would need any further permission.
(The concern was that if the code in [POptUS/BenDFO](https://github.com/POptUS/BenDFO) was derived from CUTEr _code_, then any derived code may need to be LGPL-licensed per the terms of the LGPL.)

wildsm commented:
@mdhaber, correct: the BenDFO problems are based on mathematical problems in cuter (now cutest: https://github.com/ralna/CUTEst) but were new implementations of those algebraic problems.

mdhaber commented:
That's great to hear. Thanks @wildsm! I really appreciate the work!
The [comment above](https://github.com/scipy/scipy/pull/14579#issuecomment-1274099897) suggested that the optimal values of the objective function were all zero, but sounded unsure. Is that right? (Another thing I'd like to track here is the difference in the achieved objective function value from the reference.)

wildsm commented:
> That's great to hear. Thanks @wildsm! I really appreciate the work!
> 
> The [comment above](https://github.com/scipy/scipy/pull/14579#issuecomment-1274099897) suggested that the optimal values of the objective function were zero for all problems, but sounded unsure. Is that right? (Another thing I'd like to track here is the difference in the achieved objective function value from the reference.)
>

It is **not** the case the all of the original 53(*4 since there are 4 problem types in calfun) problems admit a solution f*=0. 
Noting an achievable objection function value is a good aspiration, but probably most meaningful for the 'smooth' problem class. 
Instead, since these are sums of squares, for now it would be safest to say that the objective function is always bounded below by 0 (and hence achieving a near-zero value is a good target).
 
> As a side question - based on your experience, are there free, permissively-licensed optimizers that we don't already include in SciPy that we should add? ([Here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html)'s our current list.) We've also considered allowing `minimize` to use IPOPT (via [cyipopt](https://github.com/mechmotum/cyipopt)) as an optional dependency, which the user would need to install separately because its license is incompatible. Do you think that would be valuable?

There are several permissively licensed optimizers that could be added, but I do not want to distract from the current PR. 

If there is an issue or project that you could link me to or otherwise have a way for the broader teams to address (and potentially improve optimizer options in scipy), I will try to help.

Thanks to @sarafridov and you all for your efforts.

wildsm reviewed (commented):
I suggest changing instances of 
http://www.mcs.anl.gov/~more/dfo/
to 
https://github.com/POptUS/BenDFO
in order to clearly reflect license and updates.

sarafridov commented:
Thanks for the quick replies and info @wildsm!

mdhaber commented:
I'll make that change and merge this. I decided that the format of the benchmarks is OK for now - the system used by the global optimizers (which was used in this PR) is not standard (e.g. it uses `track_all` for timing instead of `time_all`), but there are some advantages, so there's no need to refactor.

mdhaber commented:
I ran this locally and these seem like a useful set of problems. They are quite fast to run, but they do expose some unusual behavior in some optimizers (e.g. `COBLYLA` quite frequently maxes out at 1000 function evaluations). I haven't checked that they are translated exactly, but the solvers seem to agree on the optimal value of the objective function in most cases. One oddball is 17:
```
[100.00%] ··· ============================== ============== =========== ========================
              --                                                        result type
              --------------------------------------------- ------------------------------------
               DFO benchmark problem number      solver      mean_nfev          min_obj
              ============================== ============== =========== ========================
                            17                   COBYLA         48.0       7182338.049392037
                            17                   SLSQP         139.0       1417868821.8110585
                            17                   Powell        296.0       106966.87457479208
                            17                nelder-mead      600.0       6409.119264481873
                            17                  L-BFGS-B        16.0       6977375.029202724
                            17                    BFGS         1852.0      105.50823442271016
                            17                trust-constr     2484.0      27559.144777426165
```
So it might be worth a look at the translation - did it turn into an unbounded maximization, or are the solvers really struggling?  I should check that the failure of the solvers is being tracked, too. (separately).

I'll go ahead and merge if CircleCI comes back green.

<details>

```
[100.00%] ··· ============================== ============== =========== ========================
              --                                                        result type
              --------------------------------------------- ------------------------------------
               DFO benchmark problem number      solver      mean_nfev          min_obj
              ============================== ============== =========== ========================
                            0                    COBYLA        182.0       36.00000001629636
                            0                    SLSQP          21.0       35.99999999999999
                            0                    Powell        231.0       35.99999999999998
                            0                 nelder-mead      1390.0      36.000000004916636
                            0                   L-BFGS-B        30.0       36.00000000029088
                            0                     BFGS          40.0       36.00000000000747
                            0                 trust-constr      30.0       36.00000000006579
                            1                    COBYLA        239.0       36.00000002161123
                            1                    SLSQP          21.0              36.0
                            1                    Powell        374.0       35.99999999999997
                            1                 nelder-mead      1096.0      36.00000001672908
                            1                   L-BFGS-B        50.0       36.00000000000113
                            1                     BFGS          60.0       36.00000000512027
                            1                 trust-constr      40.0        36.0000000000484
                            2                    COBYLA         93.0        8.38460155098156
                            2                    SLSQP          23.0       8.380281690140846
                            2                    Powell        217.0       8.380281690140844
                            2                 nelder-mead      276.0       8.380314322959702
                            2                   L-BFGS-B        32.0       8.380281690152533
                            2                     BFGS         268.0       8.380281698083904
                            2                 trust-constr      40.0       8.380281690145555
                            3                    COBYLA        107.0       8.400780446868142
                            3                    SLSQP          23.0       8.380284602260867
                            3                    Powell        196.0       8.380281690140844
                            3                 nelder-mead      320.0       8.380314069260285
                            3                   L-BFGS-B        40.0       8.380281690152607
                            3                     BFGS         380.0       8.380281690223411
                            3                 trust-constr      48.0       8.380281690221938
                            4                    COBYLA        103.0        9.89172925972701
                            4                    SLSQP          23.0       9.880597014925373
                            4                    Powell        256.0       9.880597014925371
                            4                 nelder-mead      267.0       9.880630904843933
                            4                   L-BFGS-B       112.0       9.880597014925657
                            4                     BFGS         462.0        9.88059701530113
                            4                 trust-constr     136.0       9.880597014929531
                            5                    COBYLA        124.0       9.880673204658704
                            5                    SLSQP          23.0       9.880597015430164
                            5                    Powell        186.0       9.880597014925371
                            5                 nelder-mead      310.0       9.880645653536046
                            5                   L-BFGS-B        40.0       9.880597014932858
                            5                     BFGS         316.0       9.880597014940241
                            5                 trust-constr     200.0       9.880597014935532
                            6                    COBYLA        1000.0      0.2761343670249867
                            6                    SLSQP         109.0     3.534301293924672e-05
                            6                    Powell        665.0     1.7404243721438573e-26
                            6                 nelder-mead      159.0     8.177661197416674e-10
                            6                   L-BFGS-B       126.0     3.0327590958858587e-07
                            6                     BFGS         114.0     9.730195046515482e-11
                            6                 trust-constr     177.0     5.200784532767023e-09
                            7                    COBYLA        1000.0      18.15444498891401
                            7                    SLSQP         295.0     2.4772818584112955e-05
                            7                    Powell        1039.0    1.1093356479670479e-31
                            7                 nelder-mead      325.0     4.3387935016943286e-10
                            7                   L-BFGS-B        33.0       18.35859455706266
                            7                     BFGS         417.0     2.2882524907281913e-11
                            7                 trust-constr     375.0     2.746553712615791e-09
                            8                    COBYLA        1000.0    0.0036908922090317185
                            8                    SLSQP         112.0     2.467526280245205e-06
                            8                    Powell         67.0              0.0
                            8                 nelder-mead      142.0     0.0003575878654431348
                            8                   L-BFGS-B       120.0     1.2590925940553836e-08
                            8                     BFGS         136.0      3.64407093097148e-12
                            8                 trust-constr     128.0     4.6652773496330196e-12
                            9                    COBYLA        1000.0      1.5616155551517263
                            9                    SLSQP         114.0     3.1446883905430233e-06
                            9                    Powell         60.0              0.0
                            9                 nelder-mead      110.0     0.0001185672468711772
                            9                   L-BFGS-B       140.0     1.0985609971404184e-05
                            9                     BFGS         160.0     4.353936427472428e-12
                            9                 trust-constr     172.0     1.1084157363858046e-10
                            10                   COBYLA        1000.0    0.0018267758852006735
                            10                   SLSQP          99.0     9.231453061284653e-05
                            10                   Powell        997.0     1.692602785645291e-13
                            10                nelder-mead      305.0     1.3905860499424258e-06
                            10                  L-BFGS-B       100.0     3.5701552391704273e-05
                            10                    BFGS         165.0     3.266272925189801e-07
                            10                trust-constr     190.0     3.4240610197075264e-07
                            11                   COBYLA        1000.0     0.007588035155492048
                            11                   SLSQP         179.0      3.44391715908274e-05
                            11                   Powell        1742.0    3.0130545641301406e-22
                            11                nelder-mead      380.0     1.2207288155937388e-06
                            11                  L-BFGS-B       165.0     2.5695461782917457e-05
                            11                    BFGS         280.0     1.908979241735041e-07
                            11                trust-constr     355.0     2.208504660413736e-07
                            12                   COBYLA        1000.0      63.35600768650094
                            12                   SLSQP          28.0        48.9842541402506
                            12                   Powell        130.0       48.98425367924135
                            12                nelder-mead      120.0       48.98425367981376
                            12                  L-BFGS-B        54.0       48.98425585246278
                            12                    BFGS          27.0       48.98425367924929
                            12                trust-constr      60.0       48.98425367924165
                            13                   COBYLA        1000.0       52.5180366249726
                            13                   SLSQP          47.0       48.984285104912686
                            13                   Powell        255.0       48.984253679248496
                            13                nelder-mead      119.0     1.7922698821282258e-09
                            13                  L-BFGS-B        90.0        48.9843135995884
                            13                    BFGS          75.0       48.984253679241974
                            13                trust-constr      84.0       48.98425367938097
                            14                   COBYLA        1000.0     0.010846968025621654
                            14                   SLSQP          36.0      0.010926876593972496
                            14                   Powell        482.0      0.008214880387696202
                            14                nelder-mead      226.0      0.008214877316410097
                            14                  L-BFGS-B        36.0      0.009974372219517814
                            14                    BFGS          92.0      0.008214877322486921
                            14                trust-constr      96.0      0.00821493131546288
                            15                   COBYLA        1000.0       8.3527643805625
                            15                   SLSQP          80.0      0.010852444332488186
                            15                   Powell        232.0       1.7749269756254313
                            15                nelder-mead      346.0      0.008214877318251797
                            15                  L-BFGS-B        92.0      0.010443321669298032
                            15                    BFGS         136.0      0.008214877306853469
                            15                trust-constr     1924.0     0.008214941209108827
                            16                   COBYLA        1000.0    0.00038308147268715565
                            16                   SLSQP          22.0     0.0005011638818990982
                            16                   Powell        570.0     0.00030750560950841397
                            16                nelder-mead      260.0     0.00030750561104950664
                            16                  L-BFGS-B        45.0     0.0004998454312085443
                            16                    BFGS         165.0     0.0003075071717883809
                            16                trust-constr     170.0     0.00030985544885250914
                            17                   COBYLA         48.0       7182338.049392037
                            17                   SLSQP         139.0       1417868821.8110585
                            17                   Powell        296.0       106966.87457479208
                            17                nelder-mead      600.0       6409.119264481873
                            17                  L-BFGS-B        16.0       6977375.029202724
                            17                    BFGS         1852.0      105.50823442271016
                            17                trust-constr     2484.0      27559.144777426165
                            18                   COBYLA        1000.0     0.010975991274795866
                            18                   SLSQP         145.0     0.0031019739183846718
                            18                   Powell        1926.0    0.0022876970870087574
                            18                nelder-mead      1200.0     0.004794624522076027
                            18                  L-BFGS-B       133.0      0.009260103353715464
                            18                    BFGS         266.0      0.002287670084218029
                            18                trust-constr     371.0      0.002287793309747557
                            19                   COBYLA        1000.0     0.01449258959907267
                            19                   SLSQP         334.0      0.002387635461912692
                            19                   Powell        2028.0     0.017136978287569753
                            19                nelder-mead      1200.0     0.11412012392582763
                            19                  L-BFGS-B       224.0      0.010392828812354508
                            19                    BFGS         518.0      0.002287670118424779
                            19                trust-constr     546.0     0.0022896826830664527
                            20                   COBYLA        1000.0     0.012314747076097223
                            20                   SLSQP         435.0     3.7403380991915257e-05
                            20                   Powell        3856.0    0.00013999662219714502
                            20                nelder-mead      1275.0    0.0037644989112963437
                            20                  L-BFGS-B       270.0     0.0028719244514205464
                            20                    BFGS         410.0     3.797654976888867e-05
                            20                trust-constr     540.0     3.911398406372046e-05
                            21                   COBYLA        1000.0     0.19021806725341522
                            21                   SLSQP         975.0      0.005234314850263207
                            21                   Powell        5562.0     0.10605836851573695
                            21                nelder-mead      1800.0      0.5377026960946113
                            21                  L-BFGS-B       500.0     0.00028813027873123817
                            21                    BFGS         1010.0    1.8827451845615612e-06
                            21                trust-constr     1900.0    4.185790050233788e-05
                            22                   COBYLA        1000.0     0.10003177690766717
                            22                   SLSQP         613.0      0.007622482850835413
                            22                   Powell        9933.0    0.0009381552314367413
                            22                nelder-mead      2400.0     0.008256161684722775
                            22                  L-BFGS-B       741.0      0.000314364893841669
                            22                    BFGS         1001.0    1.8916384785289867e-06
                            22                trust-constr     1118.0    1.3321351943695367e-05
                            23                   COBYLA        1000.0      0.2362971492708929
                            23                   SLSQP         1331.0     0.10106796888625674
                            23                   Powell        5265.0      6.567880194260053
                            23                nelder-mead      2400.0      5.580001849914321
                            23                  L-BFGS-B       806.0      0.011298205825883999
                            23                    BFGS         2002.0    0.0003297859042824798
                            23                trust-constr     1950.0    1.3757805796874542e-05
                            24                   COBYLA        1000.0    0.0027210848918468693
                            24                   SLSQP          85.0      0.039023423059962446
                            24                   Powell        1522.0      0.0758318699160596
                            24                nelder-mead      480.0      0.07558874075499751
                            24                  L-BFGS-B       100.0     0.00023411051762240796
                            24                    BFGS         108.0     4.177123227644773e-09
                            24                trust-constr     132.0     2.2154751422553546e-06
                            25                   COBYLA         32.0       256.1647690452771
                            25                   SLSQP          16.0             2020.0
                            25                   Powell        329.0       124.3621843304955
                            25                nelder-mead       72.0       124.3621848396066
                            25                  L-BFGS-B        72.0       214.29894954503868
                            25                    BFGS         147.0       124.36218235561681
                            25                trust-constr     189.0       124.36218235561688
                            26                   COBYLA        678.0       85822.20465825836
                            26                   SLSQP         130.0       85822.20163050466
                            26                   Powell        301.0       85822.24229503317
                            26                nelder-mead      333.0       85822.20162974804
                            26                  L-BFGS-B        70.0       85822.28180418117
                            26                    BFGS         215.0       85822.20162635631
                            26                trust-constr     235.0        85822.2016263593
                            27                   COBYLA        928.0       85822.20748496873
                            27                   SLSQP         252.0       85822.20162647964
                            27                   Powell        286.0       85822.29989083631
                            27                nelder-mead      397.0       85822.20164251165
                            27                  L-BFGS-B       155.0       85822.21347725767
                            27                    BFGS         591.0       85822.20162635631
                            27                trust-constr     405.0       85822.20162635841
                            28                   COBYLA        274.0     4.567933585235919e-06
                            28                   SLSQP          73.0     0.0001275291143854763
                            28                   Powell        2192.0     2.62902981284408e-22
                            28                nelder-mead      630.0     4.168900076322192e-09
                            28                  L-BFGS-B       112.0     9.673403808567107e-07
                            28                    BFGS         140.0     1.3393622123502154e-10
                            28                trust-constr     175.0     1.819529612033121e-10
                            29                   COBYLA        169.0     7.228441475939319e-07
                            29                   SLSQP          58.0     0.00017792262972978623
                            29                   Powell        2227.0    6.554533948298135e-23
                            29                nelder-mead      472.0     4.392761093121637e-07
                            29                  L-BFGS-B       112.0     2.0258935263141074e-05
                            29                    BFGS         176.0     6.136941614799513e-11
                            29                trust-constr     152.0     2.4623717975095197e-10
                            30                   COBYLA        1000.0    0.0036544588617460697
                            30                   SLSQP          72.0      0.006287589283329275
                            30                   Powell        3018.0    0.0035168774842111537
                            30                nelder-mead      813.0     0.0037430185611118747
                            30                  L-BFGS-B       180.0     0.0036331210832657988
                            30                    BFGS         270.0     0.0035168737463856397
                            30                trust-constr     333.0     0.0035168739464551393
                            31                   COBYLA        507.0     9.563260675050992e-06
                            31                   SLSQP         113.0     7.559376195674753e-05
                            31                   Powell        5182.0    1.4405961280325035e-21
                            31                nelder-mead      1008.0    1.993137661280346e-05
                            31                  L-BFGS-B       190.0     0.00016452253916588005
                            31                    BFGS         290.0     4.226424638113484e-11
                            31                trust-constr    10000.0     0.00860514728003039
                            32                   COBYLA        1000.0     0.004996093419091057
                            32                   SLSQP          76.0      0.00748298302440139
                            32                   Powell        2362.0     0.00654899679530417
                            32                nelder-mead      1142.0     0.004782384226390245
                            32                  L-BFGS-B       143.0      0.00751254641713398
                            32                    BFGS         418.0      0.006503955074807308
                            32                trust-constr    11000.0     0.03376312998006939
                            33                   COBYLA        1000.0     0.004790522744041185
                            33                   SLSQP          80.0      0.006136919804702198
                            33                   Powell        4848.0    0.0027998978593152275
                            33                nelder-mead      1799.0    0.0031707681652494613
                            33                  L-BFGS-B       312.0      0.002823158486757189
                            33                    BFGS         504.0     0.0027997618778556575
                            33                trust-constr    12000.0     0.026740538342589985
                            34                   COBYLA        1000.0     0.002477075965308623
                            34                   SLSQP          47.0     1.8202225661529484e-05
                            34                   Powell        3303.0    1.3549012557932144e-22
                            34                nelder-mead      2000.0      0.3406697389548668
                            34                  L-BFGS-B        55.0     2.991968759920641e-05
                            34                    BFGS         132.0     1.8709798615677602e-13
                            34                trust-constr     154.0     1.7082950861593002e-08
                            35                   COBYLA        1000.0      2.7742441299951137
                            35                   SLSQP          74.0       1.1056629223458196
                            35                   Powell        712.0      0.050641072866261495
                            35                nelder-mead      747.0      5.46489474798272e-05
                            35                  L-BFGS-B       108.0       1.105965377496062
                            35                    BFGS         132.0       0.6345415441186326
                            35                trust-constr     192.0       1.105960125097165
                            36                   COBYLA        1000.0     0.29941668888383516
                            36                   SLSQP         426.0      0.04140439863979013
                            36                   Powell        3474.0     0.04014330272700565
                            36                nelder-mead      2200.0     0.07609312836138543
                            36                  L-BFGS-B       684.0      0.04203767275972611
                            36                    BFGS         756.0      0.04013773868180521
                            36                trust-constr     1056.0      0.0401377475379302
                            37                   COBYLA        252.0       1.7898162978212655
                            37                   SLSQP         228.0       1.789825242674377
                            37                   Powell        682.0       1.7898135868820384
                            37                nelder-mead      809.0       1.7898135880745636
                            37                  L-BFGS-B       396.0       1.7898136280593107
                            37                    BFGS         252.0       1.7898135868811282
                            37                trust-constr     960.0       1.7898135869083511
                            38                   COBYLA        289.0       10.238983941422939
                            38                   SLSQP         216.0       10.238994699907447
                            38                   Powell        534.0       10.238973850652153
                            38                nelder-mead      1058.0      10.238973442379468
                            38                  L-BFGS-B       135.0       10.239986315277417
                            38                    BFGS         396.0       10.23897342140297
                            38                trust-constr     405.0       10.238973421446603
                            39                   COBYLA        704.0       18.281166339083896
                            39                   SLSQP         276.0       18.281172707363083
                            39                   Powell        653.0       18.281163008134268
                            39                nelder-mead      2000.0      18.281162945170095
                            39                  L-BFGS-B       165.0       18.28315258333434
                            39                    BFGS         517.0       18.28116175363265
                            39                trust-constr     638.0       18.28116175375586
                            40                   COBYLA        858.0       22.260606435048906
                            40                   SLSQP         398.0        22.2606981682437
                            40                   Powell        737.0       22.260594329663107
                            40                nelder-mead      2200.0      22.260916512172354
                            40                  L-BFGS-B       180.0       22.263680904414326
                            40                    BFGS         564.0       22.260591734938515
                            40                trust-constr     792.0       22.260591735055833
                            41                   COBYLA        832.0       26.272777626725947
                            41                   SLSQP         367.0       26.272774470672505
                            41                   Powell        775.0       26.272777250417114
                            41                nelder-mead      2400.0      26.27279966391945
                            41                  L-BFGS-B       195.0       26.27565351873632
                            41                    BFGS         611.0       26.272766396820437
                            41                trust-constr     949.0       26.27276639696674
                            42                   COBYLA         67.0     1.3779270017854998e-06
                            42                   SLSQP          83.0     1.981125865175874e-06
                            42                   Powell        975.0     9.335632166778367e-22
                            42                nelder-mead      237.0     6.465456843517023e-07
                            42                  L-BFGS-B        72.0     4.181630242250661e-06
                            42                    BFGS         114.0     1.2474610014362838e-12
                            42                trust-constr     102.0     9.301653128175893e-13
                            43                   COBYLA        103.0     5.338633400068714e-07
                            43                   SLSQP         102.0      2.21762195166915e-05
                            43                   Powell        918.0     5.5086057784253994e-21
                            43                nelder-mead      335.0      4.52708455882542e-07
                            43                  L-BFGS-B        91.0     4.4966419744444954e-08
                            43                    BFGS         140.0     2.7029960247429463e-11
                            43                trust-constr     140.0     1.3440415937610982e-12
                            44                   COBYLA        141.0     1.9045502364917844e-06
                            44                   SLSQP         157.0     2.160789669234786e-06
                            44                   Powell        1196.0    3.197963593305558e-21
                            44                nelder-mead      716.0     2.604956216519312e-07
                            44                  L-BFGS-B       117.0     4.253645603685192e-07
                            44                    BFGS         216.0     1.796256672183103e-13
                            44                trust-constr     198.0     2.0894038836280167e-13
                            45                   COBYLA        149.0      0.02029939322581242
                            45                   SLSQP          89.0     1.0351674580358287e-07
                            45                   Powell        252.0     8.959222295130253e-22
                            45                nelder-mead      375.0      4.86780002255299e-05
                            45                  L-BFGS-B        42.0      2.40098232210167e-10
                            45                    BFGS         174.0     5.367397574407069e-10
                            45                trust-constr      78.0     2.454088398353555e-06
                            46                   COBYLA        1000.0      1567939963474.8965
                            46                   SLSQP          84.0     9.996599730876226e-10
                            46                   Powell        308.0     1.1774403070777338e-20
                            46                nelder-mead      465.0     8.181505338483143e-05
                            46                  L-BFGS-B        66.0     2.7791771145450177e-10
                            46                    BFGS         192.0     5.371130442041379e-10
                            46                trust-constr     114.0     2.9953810604237804e-06
                            47                   COBYLA        206.0      0.031345389476570694
                            47                   SLSQP         156.0     8.602559202539433e-10
                            47                   Powell        402.0     3.5524832063261114e-20
                            47                nelder-mead      853.0     0.00010539806608737956
                            47                  L-BFGS-B        63.0     1.7096060207533679e-09
                            47                    BFGS         405.0     8.552560226301659e-10
                            47                trust-constr      81.0     4.041706874673093e-06
                            48                   COBYLA        268.0      0.018392055570552443
                            48                   SLSQP         139.0     0.0034366694960716385
                            48                   Powell        500.0      8.06006186354541e-20
                            48                nelder-mead      1117.0    0.0002198383834780977
                            48                  L-BFGS-B        88.0     4.803613395887105e-10
                            48                    BFGS         605.0     1.0689500477414616e-09
                            48                trust-constr      99.0     4.5226625522612525e-06
                            49                   COBYLA        369.0      0.03428531323720612
                            49                   SLSQP         304.0     1.719564972450352e-06
                            49                   Powell        600.0     3.022936823467032e-19
                            49                nelder-mead      1411.0    0.00030925186830592474
                            49                  L-BFGS-B       104.0      6.51061604981082e-10
                            49                    BFGS         702.0     1.2850314898252282e-09
                            49                trust-constr     208.0     4.8793043447732755e-06
                            50                   COBYLA        1000.0      4476072203420.505
                            50                   SLSQP         174.0       16090.611617843075
                            50                   Powell        622.0     2.8815482388462136e-19
                            50                nelder-mead      2400.0       9.64238414027864
                            50                  L-BFGS-B       156.0     1.885568155721845e-09
                            50                    BFGS         831.0     1.2659752664812657e-09
                            50                trust-constr     273.0     4.878639421667073e-06
                            51                   COBYLA        1000.0     0.043713658911804086
                            51                   SLSQP         321.0      0.003924596822050221
                            51                   Powell        5094.0    9.735906314159545e-25
                            51                nelder-mead      1600.0     0.001974655275796391
                            51                  L-BFGS-B       162.0      0.018002244485516704
                            51                    BFGS         540.0     2.530043305286086e-10
                            51                trust-constr     837.0     5.2831275528446416e-08
                            52                   COBYLA        1000.0      1267974.2895068463
                            52                   SLSQP         993.0       306.97405713164454
                            52                   Powell        6497.0      12.202607453274954
                            52                nelder-mead      1600.0      223.19648568816407
                            52                  L-BFGS-B       306.0       4.536639532281335
                            52                    BFGS         777.0       4.534879777769612
                            52                trust-constr     432.0       4.544389715643134
              ============================== ============== =========== ========================

```

</details>

sarafridov commented:
Thanks everyone!
